// webux-language/index.js
// ██╗      █████╗ ███╗   ██╗ ██████╗ ██╗   ██╗ █████╗  ██████╗ ███████╗
// ██║     ██╔══██╗████╗  ██║██╔════╝ ██║   ██║██╔══██╗██╔════╝ ██╔════╝
// ██║     ███████║██╔██╗ ██║██║  ███╗██║   ██║███████║██║  ███╗█████╗
// ██║     ██╔══██║██║╚██╗██║██║   ██║██║   ██║██╔══██║██║   ██║██╔══╝
// ███████╗██║  ██║██║ ╚████║╚██████╔╝╚██████╔╝██║  ██║╚██████╔╝███████╗
// ╚══════╝╚═╝  ╚═╝╚═╝  ╚═══╝ ╚═════╝  ╚═════╝ ╚═╝  ╚═╝ ╚═════╝ ╚══════╝

/**
 * File: index.js
 * Author: Tommy Gingras
 * Date: 2019-02-22
 * License: All rights reserved Studio Webux S.E.N.C 2015-Present
 */

"use strict";

const i18n = require("i18n");
const path = require("path");
const { Webux } = require("webux-app");

const init = (app, options) => {
  app.use((req, res, next) => {
    i18n.configure({
      locales: options.availables,
      directory: path.join(options.directory),
      defaultLocale: options.default,
      autoReload: options.autoReload,
      syncFiles: options.syncFiles,
      logDebugFn: msg => Webux.log.debug(msg)
    });

    i18n.init;

    const lang = req.headers && req.headers.lang ? req.headers.lang : "en";
    i18n.setLocale(lang);

    return next();
  });
};

const getLocale = () => {
  return i18n.getLocale();
};

module.exports = {
  getLocale,
  init
};
